### DISCUSSION TITLE
How to execute shell command from js hooks?

### DISCUSSION THREAD

**Discussion by cesboa** [pocketbase/pocketbase#3506] [Q&A] [answered]
I can't figure out how to execute a shell script, like the one from the JSVM ref example.
https://pocketbase.io/jsvm/functions/_os.exec.html

It returns the structure, but it does not execute the command. How do I execute the structured command in the terminal / shell?

Cheers,

**Comment by ganigeorgiev** [comment, marked as answer]
The `$os.exec()` function is a binding for the Go [`exec.Command`](https://pkg.go.dev/os/exec#example-Command), meaning that you should be able to call something like `$os.exec("yourCommand").run()` (_I'll consider maybe rename/alias it to `$os.cmd()` to avoid the confusion_).

**Comment by cesboa** [comment]
That works, thanks a lot!!! :)
I'm trying to get the result of the exec.run(). I have looked at the go exec.Command, but i can't figure it out :)
I have tried await, than etc. on run(), but that results in undefined.
Is there a way to get the output from the console, for like ls -sl or echo hello?

**Reply by ganigeorgiev** [reply]
You can try with [`$os.exec(...).output()`](https://pocketbase.io/jsvm/interfaces/exec.Cmd.html#output).

**Reply by cesboa** [reply]
Damn you are fast :)

That gives some output, but it looks strange.

const cmd = $os.exec('date').output()
console.log(cmd)
gives: 70,114,105,32,79,99,116,32,49,51,32,49,52,58,49,55,58,53,54,32,67,69,83,84,32,50,48,50,51,10

ls -sl gives:
116,111,116,97,108,32,57,56,54,49,54,10,32,32,32,32,48,32,100,114,119,120,114,45,120,114,45,120,32,32,32,53,32,99,101,97,32,32,115,116,97,102,102,32,32,32,32,32,32,32,49,54,48,32,79,99,116,32,49,50,32,50,51,58,49,49,32,95,83,79,85,82,67,69,10,32,32,32,32,48,32,100,114,119,120,114,45,120,114,119,120,32,32,32,53,32,99,101,97,32,32,115,116,97,102,102,32,32,32,32,32,32,32,49,54,48,32,79,99,116,32,32,49,32,48,48,58,53,51,32,100,97,116,97,10,32,32,32,32,48,32,100,114,119,120,114,45,120,114,119,120,32,32,49,50,32,99,101,97,32,32,115,116,97,102,102,32,32,32,32,32,32,32,51,56,52,32,79,99,116,32,49,51,32,49,52,58,49,56,32,112,98,95,100,97,116,97,10,32,32,32,32,48,32,100,114,119,120,114,45,120,114,119,120,32,32,32,51,32,99,101,97,32,32,115,116,97,102,102,32,32,32,32,32,32,32,32,57,54,32,83,101,112,32,32,54,32,50,51,58,50,54,32,112,98,95,104,111,111,107,115,10,57,56,54,49,54,32,45,114,119,120,114,45,120,114,45,120,64,32,32,49,32,99,101,97,32,32,115,116,97,102,102,32,32,53,48,52,56,57,52,52,48,32,79,99,116,32,32,56,32,50,50,58,53,56,32,112,111,99,107,101,116,98,97,115,101,10

**Reply by cesboa** [reply]
Aaah, sorry. it's ascii. I just have to convert it to string.
Thank's a lot for your time!

**Reply by ganigeorgiev** [reply]
Yes, the Go method returns `[]byte` (_this weekend I'll see if the auto generated JSVM docs can be easily updated to properly reflect that_).

To convert it to a plain string you can try with `String.fromCharCode(...outputArr)`.

**Reply by cesboa** [reply]
Yes, this gives me the expected output from terminal. :) Have a great weekend!

```
const cmd = String.fromCharCode(... $os.exec('ls', '-sl').output())
console.log(cmd)
```
